pkgs/by-name/ni/niri/package.nix

```
@@ -38,56 +47,76 @@ rustPlatform.buildRustPackage rec {
     };
   };
 
+  strictDeps = true;
+
   nativeBuildInputs = [
+    clang
     pkg-config
     rustPlatform.bindgenHook
-    autoPatchelfHook
-    clang
-  ];
-
-  buildInputs = [
-    wayland
-    systemd # For libudev
-    seatd # For libseat
-    libdisplay-info
-    libxkbcommon
-    libinput
-    mesa # For libgbm
-    fontconfig
-    stdenv.cc.cc.lib
-    pipewire
-    pango
   ];
 
-  runtimeDependencies = [
-    wayland
-    mesa
-    libglvnd # For libEGL
-  ];
+  buildInputs =
+    [
+      libdisplay-info
+      libglvnd # For libEGL
+      libinput
+      libxkbcommon
+      mesa # For libgbm
+      pango
+      seatd
+      wayland # For libwayland-client
+    ]
+    ++ lib.optional (withDbus || withScreencastSupport || withSystemd) dbus
+    ++ lib.optional withScreencastSupport pipewire
+    ++ lib.optional withSystemd systemd; # Also includes libudev
 
-  passthru.providedSessions = [ "niri" ];
+  buildFeatures =
+    lib.optional withDbus "dbus"
+    ++ lib.optional withDinit "dinit"
+    ++ lib.optional withScreencastSupport "xdp-gnome-screencast"
+    ++ lib.optional withSystemd "systemd";
+  buildNoDefaultFeatures = true;
 
-  postPatch = ''
-    patchShebangs ./resources/niri-session
-    substituteInPlace ./resources/niri.service \
-      --replace-fail '/usr/bin' "$out/bin"
-  '';
+  postInstall =
+    ''
+      install -Dm0644 resources/niri.desktop -t $out/share/wayland-sessions
+    ''
+    + lib.optionalString withSystemd ''
+      install -Dm0755 resources/niri-session -t $out/bin
+      install -Dm0644 resources/niri{-shutdown.target,.service} -t $out/share/systemd/user
+      install -Dm0644 resources/niri-portals.conf -t $out/share/xdg-desktop-portal
+    '';
 
-  postInstall = ''
-    install -Dm0755 ./resources/niri-session -t $out/bin
-    install -Dm0644 resources/niri.desktop -t $out/share/wayland-sessions
-    install -Dm0644 resources/niri-portals.conf -t $out/share/xdg-desktop-portal
-    install -Dm0644 resources/niri{-shutdown.target,.service} -t $out/share/systemd/user
-  '';
+  env = {
+    # Force linking with libEGL and libwayland-client
+    # so they can be discovered by `dlopen()`
+    RUSTFLAGS = toString (
```

`toString` of an array? i actually had to test this in `nix repl` now because that looks like it shouldn't be allowed. and indeed, `toString (["a" "b"]) == "a b"`. that's... very unintuitive? please no

```suggestion
    RUSTFLAGS = builtins.concatStringsSep " " (
```

---

This is [well documented](https://nix.dev/manual/nix/2.24/language/builtins#builtins-toString). It is also overwhelmingly common in nixpkgs, with [192+ files](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22toString+%5B%22+path%3A%2F%5Epkgs%5C%2F%2F&type=code&p=3) doing the same thing compared to [only 4](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22builtins.concatStringsSep+%5C%22+%5C%22+%5B%22+path%3A%2F%5Epkgs%5C%2F%2F&type=code) following this suggestion (or [~34](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22builtins.concatStringsSep+%5C%22+%5C%22%22+path%3A%2F%5Epkgs%5C%2F%2F&type=code) if we're a more vague with the search). I don't see any reason to reimplement the functionality of `toString` when this is the case, as that would be a lot more unintuitive

---

I agree with @sodiboo on this one.

---

`toString` of an list shouldn't be unintuitive, that's how `buildInputs, checkInputs....` and others work. interesting that `env` argument doesn't allow list. 

https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix#L584
 `(isString v || isBool v || isInt v || isDerivation v)` but since it allows derivation, my guess list should be valid too

---

I guess the real question probably is, whether we really want to go into this discussion, or whether we can just add the more explicit `concatStringsSep` and get this PR merged.

---

I don't mind either way. I am just adding more info in support of @getchoo 
eg https://nix.dev/manual/nix/2.24/language/derivations 

```
Lists of the previous types are also allowed. They are simply concatenated, separated by spaces.
```

https://github.com/NixOS/nixpkgs/tree/master/maintainers#definition-and-role-of-the-maintainer
as existing maintainer, @sodiboo request should be honored, but we can still have a discussion about choices, pros/cons

---

> or whether we can just add the more explicit concatStringsSep and get this PR merged

I'm a bit confused as to why this would be a blocker? These are functionally identical, and as noted [in the contributing guide](https://github.com/NixOS/nixpkgs/blob/b53cf65876d5b5dd6fc8d0df3f3f3f185fee164a/CONTRIBUTING.md#how-to-review-pull-requests):

> If you find anything related that could be improved but is not immediately required for acceptance [...] please remember not to make such additional considerations a blocker

If anyone believes there are problems with this practice and would like to discuss it, I'd recommend taking the guide's suggestion of opening a tracking issue -- especially with how it seems `toString` is the default tree-wide

---

> This is [well documented](https://nix.dev/manual/nix/2.24/language/builtins#builtins-toString).

I don't agree that this makes it okay. One line down you'll also see that `false` and `null` map to the same string value, and `true -> "1"`, which i also think is unintuitive along with the quirks of [passing a path to it](https://github.com/NixOS/nix/issues/7338). Language design mistakes happen, [Nix has a history of them](https://nix.dev/guides/best-practices), and they're hard to meaningfully change or deprecate, but this doesn't mean we should rely on their behaviour and pretend like that's obvious. To be clear, the issue isn't that this is an obscure undocumented quirk of Nix, it's that i don't hesitate to even call it a quirk, because what `toString` means in this context is not very obvious, and i had to look up what it even means to know what it did here.

> [...] reimplement the functionality of `toString` [...], as that would be a lot more unintuitive

unintuitive? to read `builtins.concatStringsSep " "` over `toString`?

it's also not reimplementing anything, it's just... calling a different builtin.

> `toString` of an list shouldn't be unintuitive, that's how `buildInputs`, `checkInputs`.... and others work.

No? These take lists. You don't have to call `toString` on them. If `mkDerivation` uses `toString` to concatenate them under the hood that's their problem, but this behaviour of `toString` is not obvious just from how these other things work, because you aren't expected to concatenate the lists or know what format they should be concatenated in.

> I guess the real question probably is, whether we really want to go into this discussion, [...]

lol too late

> I'm a bit confused as to why this would be a blocker?

I guess i didn't mean to make this a Blocker™. Like, i don't really care *that much* and this is a trivial change with zero real effect. I didn't mean to start some discussion about this, because i genuinely thought this was a very unobjectionable suggestion, but it seems i had misjudged the case. Without this suggestion, everything still works and behaves identically, so whatever, go ahead and merge it, the actual package is fine.

> If anyone believes there are problems with <the practice of using `toString` to space-concat lists> and would like to discuss it, [...]

I do think there are problems with it, but honestly i don't care enough to drive much more meaningful discussion about the issue other than the single paragraph i posted at the start of this comment. There's *a lot* of stuff that bugs me about Nix, and this behaviour of `toString` is just a tiny drop in that ocean which ultimately matters very little.

---

@sodiboo greetings !

> 
> This is [well documented](https://nix.dev/manual/nix/2.24/language/builtins#builtins-toString).
> `I don't agree that this makes it okay` 

this is ok, but should not affect this PR

as it has been pointed multiple times, this is well expected behavior can we agree to proceed with this PR ?